File: kakao.app/js/ked/edit/complete.js
var _k_ = {empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }, min: function () { var m = Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.min.apply(_k_.min,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n < m ? n : m}}}; return m }, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}}

var complete

import kxk from "../../kxk.js"
let kseg = kxk.kseg
let kutil = kxk.kutil
let post = kxk.post

import theme from "../theme/theme.js"

import belt from "./tool/belt.js"

import specs from "./specs.js"


complete = (function ()
{
    function complete (editor)
    {
        this.editor = editor
    
        this["onChoicesAction"] = this["onChoicesAction"].bind(this)
        this["onWheel"] = this["onWheel"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["onEditorLayout"] = this["onEditorLayout"].bind(this)
        this.name = this.editor.name + '_complete'
        this.choices = new choices_class(this.editor.cells.screen,`${this.name}_choices`,['scrllr'])
        this.choices.focusable = false
        this.choices.scroll.handle = '▐'
        this.color = {complete:theme.editor_selection}
        this.choices.setColor('bg',theme.editor_complete_choices)
        this.choices.scroll.setColor('bg',theme.editor_complete_choices)
        this.choices.scroll.setColor('knob',theme.editor_complete_choices_scroll)
        this.choices.scroll.setColor('dot',theme.editor_complete_choices_scroll)
        this.choices.on('action',this.onChoicesAction)
        this.visible = false
    }

    complete.prototype["complete"] = function ()
    {
        var after, before, hcw, tct, tcw, turd

        before = this.editor.state.chunkBeforeCursor()
        after = this.editor.state.chunkAfterCursor()
        tcw = kseg.tailCountWord(before)
        hcw = kseg.headCountWord(after)
        if (tcw && hcw)
        {
            return
        }
        tct = kseg.tailCountTurd(before)
        turd = before
        if (tct)
        {
            turd = before.slice(before.length - 1)
        }
        else if (tcw && tcw < before.length)
        {
            turd = before.slice(before.length - tcw)
        }
        return this.word(turd)
    }

    complete.prototype["word"] = function (turd)
    {
        this.turd = turd
    
        var before, ch, ci, cx, cy, h, head, inserts, mc, mlw, x, y

        if (_k_.empty(this.turd))
        {
            this.hide()
            return
        }
        before = this.editor.state.chunkBeforeCursor()
        this.words = kseg.chunks(this.editor.state.s.lines).map(function (chunk)
        {
            return chunk.chunk
        })
        this.words = belt.prepareWordsForCompletion(before,this.turd,this.words)
        if (inserts = specs.trigger[this.turd])
        {
            if (this.turd === '>')
            {
                if (before === '>')
                {
                    this.words = inserts.concat(this.words)
                }
            }
            else
            {
                this.words = inserts.concat(this.words)
            }
        }
        this.visible = !_k_.empty(this.words)
        if (_k_.empty(this.words))
        {
            return
        }
        mc = this.editor.state.mainCursor()
        head = this.words[0]
        cx = mc[0] - this.editor.state.s.view[0]
        cy = mc[1] - this.editor.state.s.view[1]
        var list = _k_.list(head.slice(this.turd.length))
        for (ci = 0; ci < list.length; ci++)
        {
            ch = list[ci]
            this.editor.cells.set(cx + ci,cy,ch,'#fff',theme.selection)
        }
        if (this.words.length <= 1)
        {
            return this.choices.clear()
        }
        else
        {
            mlw = _k_.max(1,belt.widthOfLinesIncludingColorBubbles(this.words))
            h = _k_.min(8,this.words.length)
            x = this.editor.cells.x + cx - this.turd.length
            y = this.editor.cells.y + cy + 1
            this.choices.layout(x,y,mlw + 3,h)
            this.choices.set(this.words.map(function (w)
            {
                return ' ' + w
            }))
            return this.choices.selectFirst()
        }
    }

    complete.prototype["hide"] = function ()
    {
        if (!this.visible)
        {
            return
        }
        this.editor.state.syntax.setSegls(this.editor.state.s.lines)
        return this.visible = false
    }

    complete.prototype["hidden"] = function ()
    {
        return !this.visible
    }

    complete.prototype["onEditorLayout"] = function ()
    {
        if (this.hidden())
        {
            return
        }
    }

    complete.prototype["handleKey"] = function (key, event)
    {
        if (this.hidden() || _k_.empty(this.words))
        {
            return 'unhandled'
        }
        switch (key)
        {
            case 'tab':
            case 'return':
                return this.apply()

            case 'esc':
                return this.hide()

            case 'up':
            case 'down':
                if (this.words.length > 1)
                {
                    return this.moveSelection(key)
                }
                break
        }

        return 'unhandled'
    }

    complete.prototype["onMouse"] = function (event)
    {
        var cret

        if (this.hidden())
        {
            return
        }
        cret = this.choices.onMouse(event)
        if (!cret && _k_.in(event.type,['press','drag','release']))
        {
            this.hide()
            return
        }
        return cret
    }

    complete.prototype["onWheel"] = function (event)
    {
        if (this.hidden())
        {
            return
        }
        return this.choices.onWheel(event)
    }

    complete.prototype["moveSelection"] = function (dir)
    {
        return this.choices.moveSelection(dir)
    }

    complete.prototype["apply"] = function ()
    {
        var word

        word = this.currentWord()
        if (specs.inserts[word])
        {
            this.editor.state.delete('back')
            this.editor.state.insert(word)
        }
        else
        {
            this.editor.state.insert(word.slice(this.turd.length))
        }
        post.emit('focus','editor')
        return this.complete()
    }

    complete.prototype["onChoicesAction"] = function (action, choice)
    {
        switch (action)
        {
            case 'click':
                return this.apply()

        }

    }

    complete.prototype["currentWord"] = function ()
    {
        var word

        word = this.choices.current({trim:'front'})
        if (_k_.empty(word))
        {
            word = this.words[0]
        }
        return word
    }

    complete.prototype["preDrawLines"] = function (lines)
    {
        var pos, word

        if (this.hidden() || _k_.empty(this.words))
        {
            return lines
        }
        word = this.currentWord()
        lines = lines.asMutable()
        var list = _k_.list(this.editor.state.s.cursors)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            pos = list[_a_]
            kutil.replace(lines[pos[1]],pos[0],0,kseg(word.slice(this.turd.length)))
        }
        this.editor.state.syntax.setSegls(kseg.segls(lines))
        return lines
    }

    complete.prototype["drawCompletion"] = function ()
    {
        var ci, cx, cy, pos, word

        if (this.hidden() || _k_.empty(this.words))
        {
            return
        }
        word = this.currentWord()
        var list = _k_.list(this.editor.state.s.cursors)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            pos = list[_a_]
            cx = pos[0] - this.editor.state.s.view[0]
            cy = pos[1] - this.editor.state.s.view[1]
            for (var _b_ = ci = 0, _c_ = word.length - this.turd.length; (_b_ <= _c_ ? ci < word.length - this.turd.length : ci > word.length - this.turd.length); (_b_ <= _c_ ? ++ci : --ci))
            {
                this.editor.cells.set_bg(cx + ci,cy,this.color.complete)
            }
        }
    }

    complete.prototype["drawPopup"] = function ()
    {
        var cx, cy, fx, fy, h, mc, w, x, y

        if (this.hidden() || _k_.empty(this.words))
        {
            return
        }
        if (this.words.length <= 1)
        {
            return
        }
        mc = this.editor.state.mainCursor()
        cx = mc[0] - this.editor.state.s.view[0]
        cy = mc[1] - this.editor.state.s.view[1]
        fx = cx - this.turd.length - 1
        x = fx + 1 + this.editor.cells.x
        y = cy + this.editor.cells.y + 2
        w = this.choices.cells.cols + 1
        h = this.choices.cells.rows
        fy = cy + 1
        this.editor.cells.draw_frame(fx,fy,fx + w + 1,fy + h + 1,{fg:theme.editor_complete_choices,bg:'#000'})
        this.choices.layout(x,y,w,h)
        return this.choices.draw()
    }

    return complete
})()

export default complete;

File: kakao.app/love/lua/edit/complete.lua
--[[
 0000000   0000000   00     00  00000000   000      00000000  000000000  00000000
000       000   000  000   000  000   000  000      000          000     000     
000       000   000  000000000  00000000   000      0000000      000     0000000 
000       000   000  000 0 000  000        000      000          000     000     
 0000000   0000000   000   000  000        0000000  00000000     000     00000000
--]]

-- use ../../kxk ▪ kseg kutil post
-- use ../theme  ◆ theme 
-- use ./tool    ◆ belt
-- use           ◆ specs

choices = require "view.menu.choices"


local complete = class("complete")
    


function complete:init(editor) 
        self.editor = editor
        
        self.name = self.editor.name .. '_complete'
        
        self.choices = choices(self.editor.cells.screen, "" .. tostring(self.name) .. "_choices", array('scrllr'))
        self.choices.focusable = false
        
        self.color = {
            bg = theme.complete.bg, 
            complete = theme.selection.span, 
            scroll = theme.complete.scroll
            }
        
        self.choices:setColor('hover', {bg = theme.hover.bg, blur = theme.hover.bg})
        self.choices:setColor('bg', self.color.bg)
        if self.choices.scroll then 
            self.choices.scroll:setColor('bg', self.color.bg)
            self.choices.scroll:setColor('knob', self.color.scroll)
            self.choices.scroll:setColor('dot', self.color.scroll)
        end
        
        self.choices:on('action', self.onChoicesAction)
        
        self.visible = false
        return self
    end

--  0000000   0000000   00     00  00000000   000      00000000  000000000  00000000  
-- 000       000   000  000   000  000   000  000      000          000     000       
-- 000       000   000  000000000  00000000   000      0000000      000     0000000   
-- 000       000   000  000 0 000  000        000      000          000     000       
--  0000000   0000000   000   000  000        0000000  00000000     000     00000000  


function complete:complete() 
        if true then 
            return
        end
        
        local before = self.editor.state:chunkBeforeCursor()
        local after = self.editor.state:chunkAfterCursor()
        
        local hcw = kseg.headCountWord(after)
        local tcw = kseg.tailCountWord(before)
        
        -- log "complete.complete #{tcw} #{hcw}"
        
        if ((tcw > 1) and hcw) then return end -- don't complete if cursor is inside a word
        
        local tct = kseg.tailCountTurd(before)
        
        local turd = before
        if tct then 
            -- turd = before[before.length-1..] # start from the last punctuation character
            turd = string.sub(before, (#before - 1)) -- start from the last punctuation character
        elseif (tcw and (tcw < #before)) then 
            -- turd = before[before.length-tcw..] # start from the last word
            turd = string.sub(before, (#before - tcw)) -- start from the last word
        end
        
        return self:word(turd)
    end

-- 000   000   0000000   00000000   0000000    
-- 000 0 000  000   000  000   000  000   000  
-- 000000000  000   000  0000000    000   000  
-- 000   000  000   000  000   000  000   000  
-- 00     00   0000000   000   000  0000000    


function complete:word(turd) 
        self.turd = turd
        
        -- log "complete.word(#{@turd}"
        
        if empty(self.turd) then 
            self:hide()
            return
        end
        
        local before = self.editor.state.chunkBeforeCursor()
        
        self.words = kseg.chunks(self.editor.state.s.lines).map(function (chunk) 
    return chunk.chunk
end)
        self.words = belt.prepareWordsForCompletion(before, self.turd, self.words)
        
        if (#self.turd == 1) then 
            local inserts = specs.trigger[self.turd] -- prepend special completions, eg ~O 
            if inserts then 
                -- log 'add specs trigger' inserts
                self.words = inserts.concat(self.words)
            end
        end
        
        self.visible = valid(self.words)
        
        if empty(self.words) then return end
        
        local mc = self.editor.state.mainCursor()
        
        local head = self.words[0]
        
        local cx = (mc[0] - self.editor.state.s.view[0])
        local cy = (mc[1] - self.editor.state.s.view[1])
        
        if (#self.words <= 1) then 
            return self.choices.clear()
        else 
            local mlw = max(1, belt.widthOfLinesIncludingColorBubbles(self.words))
            
            local h = min(8, #self.words)
            local x = ((self.editor.cells.x + cx) - #self.turd)
            local y = ((self.editor.cells.y + cy) + 1)
            
            self.choices.layout(x, y, (mlw + 3), h)
            self.choices.set(self.words.map(function (w) return ' ' .. w end))
            return self.choices.selectFirst()
        end
    end

-- 000   000  000  0000000    00000000  
-- 000   000  000  000   000  000       
-- 000000000  000  000   000  0000000   
-- 000   000  000  000   000  000       
-- 000   000  000  0000000    00000000  


function complete:hide() 
        if not self.visible then return end
        self.editor.state.syntax.setSegls(self.editor.state.s.lines)
        self.visible = false
        return self.visible
    end


function complete:hidden() 
    return not self.visible
    end


function complete:onEditorLayout() 
        if self:hidden() then 
    return
           end
    end

-- ███   ███  ████████  ███   ███
-- ███  ███   ███        ███ ███ 
-- ███████    ███████     █████  
-- ███  ███   ███          ███   
-- ███   ███  ████████     ███   


function complete:handleKey(key, event) 
        if (self:hidden() or empty(self.words)) then return 'unhandled' end
        
        if (key == 'tab') or (key == 'return') then return self:apply()
        elseif (key == 'esc') then return self:hide()
        elseif (key == 'up') or (key == 'down') then 
                if (#self.words > 1) then 
                    return self:moveSelection(key)
                end
        end
        
        return 'unhandled'
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function complete:onMouse(event) 
        if self:hidden() then return end
        
        local cret = self.choices.onMouse(event)
        
        if (not(cret) and array('press', 'drag', 'release'):has(event.type)) then 
            self:hide()
            return
        end
        
        return cret
    end


function complete:onWheel(event) 
        if self:hidden() then return end
        
        return self.choices.onWheel(event)
    end

--  0000000  00000000  000      00000000   0000000  000000000  
-- 000       000       000      000       000          000     
-- 0000000   0000000   000      0000000   000          000     
--      000  000       000      000       000          000     
-- 0000000   00000000  0000000  00000000   0000000     000     


function complete:moveSelection(dir) 
        return self.choices.moveSelection(dir)
    end

--  0000000   00000000   00000000   000      000   000  
-- 000   000  000   000  000   000  000       000 000   
-- 000000000  00000000   00000000   000        00000    
-- 000   000  000        000        000         000     
-- 000   000  000        000        0000000     000     


function complete:apply() 
        local word = self:currentWord()
        
        if specs.inserts[word] then 
            self.editor.state.delete('back')
            self.editor.state.insert(word)
        else 
            -- @editor.state.insert word[@turd.length..]
            self.editor.state.insert(string.sub(word, #self.turd))
        end
        
        post:emit('focus', 'editor')
        return self:complete()
    end

--  ███████  ███   ███   ███████   ███   ███████  ████████    ███████          ███████    ███████  █████████  ███  ███   ███
-- ███       ███   ███  ███   ███  ███  ███       ███        ███              ███   ███  ███          ███     ███  ████  ███
-- ███       █████████  ███   ███  ███  ███       ███████    ███████          █████████  ███          ███          ███ █ ███
-- ███       ███   ███  ███   ███  ███  ███       ███             ███         ███   ███  ███          ███          ███  ████
--  ███████  ███   ███   ███████   ███   ███████  ████████   ███████          ███   ███   ███████     ███          ███   ███


function complete:onChoicesAction(action, choice) 
        if (action == 'click') then 
    return self:apply()
        end
    end

--  ███████  ███   ███  ████████   ████████   ████████  ███   ███  █████████      ███   ███   ███████   ████████   ███████  
-- ███       ███   ███  ███   ███  ███   ███  ███       ████  ███     ███         ███ █ ███  ███   ███  ███   ███  ███   ███
-- ███       ███   ███  ███████    ███████    ███████   ███ █ ███     ███         █████████  ███   ███  ███████    ███   ███
-- ███       ███   ███  ███   ███  ███   ███  ███       ███  ████     ███         ███   ███  ███   ███  ███   ███  ███   ███
--  ███████   ███████   ███   ███  ███   ███  ████████  ███   ███     ███         ██     ██   ███████   ███   ███  ███████  


function complete:currentWord() 
        local word = self.choices.current({trim = 'front'})
        if empty(word) then 
            word = self.words[0]
        end
        
        return word
    end


function complete:preDrawLines(lines) 
        if (self:hidden() or empty(self.words)) then return lines end
        
        local word = self:currentWord()
        lines = lines.asMutable()
        for pos in self.editor.state.s.cursors do 
            -- kutil.replace lines[pos[1]] pos[0] 0 kseg(word[@turd.length..])
            kutil.replace(lines[pos[1]], pos[0], 0, kseg(string.sub(word, #self.turd)))
        end
        
        self.editor.state.syntax.setSegls(kseg.segls(lines))
        return lines
    end

-- 0000000    00000000    0000000   000   000       0000000   0000000   00     00  00000000   000    000000000  000   000  
-- 000   000  000   000  000   000  000 0 000      000       000   000  000   000  000   000  000       000     0000  000  
-- 000   000  0000000    000000000  000000000      000       000   000  000000000  00000000   000       000     000 0 000  
-- 000   000  000   000  000   000  000   000      000       000   000  000 0 000  000        000       000     000  0000  
-- 0000000    000   000  000   000  00     00       0000000   0000000   000   000  000        0000000   000     000   000  


function complete:drawCompletion() 
        if (self:hidden() or empty(self.words)) then return end
        
        local word = self:currentWord()
        
        for pos in self.editor.state.s.cursors do 
            local cx = (pos[0] - self.editor.state.s.view[0])
            local cy = (pos[1] - self.editor.state.s.view[1])
            
            for ci = 0, (#word - #self.turd)-1 do 
                self.editor.cells.set_bg((cx + ci), cy, self.color.complete)
            end
        end
        
        return self
    end

-- 0000000    00000000    0000000   000   000       00000000    0000000   00000000   000   000  00000000   
-- 000   000  000   000  000   000  000 0 000       000   000  000   000  000   000  000   000  000   000  
-- 000   000  0000000    000000000  000000000       00000000   000   000  00000000   000   000  00000000   
-- 000   000  000   000  000   000  000   000       000        000   000  000        000   000  000        
-- 0000000    000   000  000   000  00     00       000         0000000   000         0000000   000        


function complete:drawPopup() 
        if (self:hidden() or empty(self.words)) then return end
        if (#self.words <= 1) then return end
        
        local mc = self.editor.state.mainCursor()
        
        local cx = (mc[0] - self.editor.state.s.view[0])
        local cy = (mc[1] - self.editor.state.s.view[1])
        
        local fx = ((cx - #self.turd) - 2)
        local x = ((fx + 1) + self.editor.cells.x)
        local y = ((cy + self.editor.cells.y) + 2)
        local w = (self.choices.cells.cols + 1)
        local h = self.choices.cells.rows
        
        local fy = (cy + 1)
        
        self.editor.cells.draw_rounded_border(fx, fy, ((fx + w) + 1), ((fy + h) + 1), {fg = self.color.bg})
        
        self.choices.layout(x, y, w, h)
        return self.choices.draw()
    end

return complete